src/Telemetry/AzureDevOpsTimeline/BuildLogScraper.cs

```diff
@@ -0,0 +1,59 @@
+using Castle.Core.Logging;
+using Microsoft.DotNet.Internal.AzureDevOps;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Text;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.DotNet.AzureDevOpsTimeline
+{
+    public class BuildLogScraper : IBuildLogScraper
+    {
+        private readonly ILogger<BuildLogScraper> _logger;
+        private readonly IAzureDevOpsClient _azureDevOpsClient;
+
+        public BuildLogScraper(ILogger<BuildLogScraper> logger, IAzureDevOpsClient client)
+        {
+            _azureDevOpsClient = client;
+            _logger = logger;
+        }
+
+        public Task<string> ExtractMicrosoftHostedPoolImageNameAsync(string logUri, CancellationToken cancellationToken)
+            => ExtractImageNameAsync(logUri, azurePipelinesRegex, cancellationToken);
+
+        public Task<string> ExtractOneESHostedPoolImageNameAsync(string logUri, CancellationToken cancellationToken)
+            => ExtractImageNameAsync(logUri, oneESRegex, cancellationToken);
+
+        private async Task<string> ExtractImageNameAsync(string logUri, Regex imageNameRegex, CancellationToken cancellationToken)
+        {
+            string logText = await _azureDevOpsClient.TryGetLogContents(logUri, cancellationToken);
+
+            if (string.IsNullOrEmpty(logText))
+            {
+                _logger.LogInformation($"Got empty log file for '{logUri}'");
+                return null;
+            }
+
+            Match match = imageNameRegex.Match(logText);
```

Comment:
Is there some API to get only the first Match or does this do that already? So that we're not analyzing the whole log necessarily
(asking)

Reply:
But we don't want all of them, right? We want the first?

Reply:
yes, we just care about the first one

Reply:
We also should read the log as a Stream as it can be huge and we can kill the read quite early

Reply:
Are the lines in a predictably early place?  Like, could we limit reading to, say, the first 100 lines?  So that for logs that _don't_ have the pattern, we don't spend a ton of time pointlessly searching the logs.